### DISCUSSION TITLE
new Self behavior

### DISCUSSION THREAD

**Discussion by mehdigmira** [microsoft/pyright#6615] [Q&A]
Hey, I'm trying to update pyright and seems recent version have changed a bit the way the "Self" type works. 

Now the code below shows an error `Argument of type "Sequence[B]" cannot be assigned to parameter "items" of type "Sequence[Self@B]" in function "convert"`

```python
from __future__ import annotations

from collections.abc import Sequence
from typing import Self

from typing_extensions import override


class A:
    def convert(self, items: Sequence[Self])-> Sequence[Self]:
        ...


class B(A):
    @override
    def convert(self, items: Sequence[Self])-> Sequence[Self]:
        return super().convert(get_b())


def get_b()-> Sequence[B]:
    ...

```

I've been thinking about this but I can't see while the `super().convert(get_b())` is unsafe. Self is supposed to be a shortcut for 
```python
T = TypeVar("T", boundA)
class A:
    def convert(self: T, items: Sequence[T])-> Sequence[T]:
        ...
```

So `super().convert(get_b())` seems to be legit

**Comment by erictraut** [comment]
This is unsafe because `A.convert` indicates that it accepts a `Sequence[Self]`, but you're passing a `Sequence[B]`. If the `convert` method is called from some other subclass (say, e.g. `C`, which derives from `B`), then this is not compatible.

```python
class C(B):
    pass

c = C()
c.convert([C()]) # Should return a `Sequence[C]`
```

**Reply by mehdigmira** [reply]
I'd agree if the error reported was about B.convert return type, but the error is actually about the `super().convert(get_b())` part.

So that even if B's type was 

```python 
class B(A):
    @override
    def convert(self, items: Sequence[Self])-> Sequence[Self]:
        super().convert(get_b())
        return super().convert(items)
```

pyright still complaints.

I really think super().convert() should not cause any error as long as it's called with something that subclasses A.

**Reply by erictraut** [reply]
I think pyright is correct from a type perspective, and I don't see any justification for special-casing the behavior in this particular case. I'm not even sure what makes this case special in any way.

I think we can agree that this should generate an error, right?

```python
from typing import Self

class A:
    pass

class B(A):
    def method1(self):
        x: Self = B() # Type error: `B` is not assignable to `Self@B`
```

For the same reason, it's clear this should be an error, right?

```python
from typing import Self

class A:
    def takes_self(self, x: Self):
        pass

class B(A):
    def method1(self):
        self.takes_self(B()) # Type error: `B` is not assignable to `Self@B`
```

Changing `self.takes_self` to `super().takes_self` is no different. It binds the `takes_self` method to an object of type `Self@B`, and `B` is still incompatible with `Self@B`.

```python
class B(A):
    def method1(self):
        super().takes_self(B()) # Type error: `B` is not assignable to `Self@B`
```

**Reply by mehdigmira** [reply]
IMO, your second example shouldn't log an error 
```python
from typing import Self

class A:
    def takes_self(self, x: Self):
        pass

class B(A):
    def method1(self):
        self.takes_self(B()) # Type error: `B` is not assignable to `Self@B`
```

PEP673 [claims](https://peps.python.org/pep-0673/#specification) that this code should be strictly equivalent to

```python
T = TypeVar("T", bound=A)

class A:
    def takes_self(self: T, x: T):
        pass

class B(A):
    def method1(self):
        self.takes_self(B())
```

calling `self.takes_self(B())` just means `T` should be evaluated to be equal to B. I don't see why that's unsafe.
It seems to me like pyright considers Self to be a type in itself, while PEP673 presents it as a shortcut for a TypeVar.

**Reply by erictraut** [reply]
> PEP673 [claims](https://peps.python.org/pep-0673/#specification) that this code should be strictly equivalent to...

Not quite. You forgot to include the equivalent `Self` annotation for `B.method1`.

```python
from typing import TypeVar

SelfA = TypeVar("SelfA", bound="A")

class A:
    def takes_self(self: SelfA, x: SelfA):
        pass

SelfB = TypeVar("SelfB", bound="B")

class B(A):
    def method1(self: SelfB):
        self.takes_self(B())  # Type violation
```

If you change `B.method1` to the following, then you eliminate the type violation:

```python
class B(A):
    def method1(self: "B"):
        self.takes_self(B())
```

**Reply by mehdigmira** [reply]
Thank you for taking the time to explain. I understand better how pyright considers things. I thought self was only inferred as being a typevar Self when it was explicitely mentioned in the signature.

Because 

```python
    def method1(self: SelfB):
        self.takes_self(B())
```

doesn't actually make sense from a typing point of view (and is even flagged as an error by pyright when written that way). A typevar only makes sense when it appears multiple time. But I understand the logic now.

But let's say the class B was

```python
class B(A):
    def method1(self):
        super().takes_self(B())  
```

How can I make it work ? I've tried annotating self as B, but i still have an error `Argument of type "B" cannot be assigned to parameter "x" of type "Self@B" in function "takes_self"`.
Btw I'm not trying to find a loophole, but actually trying to fix some errors that appeared since I've upgraded pyright.

**Reply by erictraut** [reply]
> I understand better how pyright considers things

This is what the typing rules dictate. Pyright is simply enforcing the typing rules here.

> How can I make it work?

It depends on the intent of your code. If `A.takes_self` requires that `self` and `x` are always instances of the same class, then this is the approach I recommend:

```python
from typing import Self

class A:
    def takes_self(self, x: Self):
        pass

class B(A):
    def method1(self):
        super().takes_self(self.__class__()) # This could also be `type(self)()` if you prefer
```

If `A.takes_self` doesn't require that `self` and `x` are always instance of the same class — but simply requires that `x` is some subclass of `A` regardless of the specialized type of `self`, then this is the approach I recommend:

```python
class A:
    def takes_self(self, x: "A"):
        pass

class B(A):
    def method1(self):
        super().takes_self(B())
```

**Reply by mehdigmira** [reply]
> This is what the typing rules dictate. Pyright is simply enforcing the typing rules here.

 Well in the same PEP, you can find the following paragraph:

> One implementation strategy is to simply desugar the former to the latter in a preprocessing step. If a method uses Self in its signature, the type of self within a method will be Self. In other cases, the type of self will remain the enclosing class.

"In other cases, the type of self will remain the enclosing class." => Pyright doesn't do this according to your message "Not quite. You forgot to include the equivalent Self annotation for B.method1." . So this is specific to Pyright, and not actually dictated by typing rules.

> It depends on the intent of your code ...


I'm in the first usecase, but i'm not intanciating B myself. The code is rather smth like

```python
from __future__ import annotations

from typing import Self


class A:
    def takes_self(self, x: Self):
        pass

class B(A):
    def method1(self):
        super().takes_self(get_b()) 


def get_b() -> B:
    ...
```

and get_b is actually not smth I have control over.

I have tried poking around this, and I'm actually surprised by some results

```python
from __future__ import annotations

from typing import Self


class A:
    def takes_self(self, x: Self)->Self:
        ...

class B(A):
    ...


reveal_type(A.takes_self(B(), B()))  # this is A
```

the revealed type is A, I would expect it to be B, if the implementation is equivalent to 

```python
SelfA = TypeVar("SelfA", bound="A")

class A:
    def takes_self(self: SelfA, x: SelfA)->SelfA:
        ...
```

I really find this new behavior has some very counter-intuitive effect  (or maybe I'm the one getting confused 😵 )

**Reply by Azureblade3808** [reply]
IMO, `typing.Self` should not be allowed in parameter types.

When `Self` is used in parameter types in a covariant or invariant way, it breaks Liskov.
```python
from typing import Self

class A:
    def takes_self(self, arg: Self) -> None:
        assert isinstance(arg, type(self))

class B(A): pass

def get_a() -> A:
    return B()

a0: A = A()  # This is an instance of A.
a1: A = get_a()  # This is actually an instance of B.
a1.takes_self(a0)  # Type-checks but breaks in runtime.
```

When `Self` is used in parameter types in a contravariant way, it is safe, but `Self` can be replaced with the concrete type of `self` in most cases.
